site/en/blog/renderingng-data-structures/index.md

```diff
@@ -0,0 +1,754 @@
+
+RenderingNG uses property trees for many purposes, including:
+
+- Separating compositing from paint, and compositing off the main thread.
+- Determining an optimal compositing / draw strategy.
+- Measuring
+[IntersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API)
+geometry.
+- Avoiding work for offscreen elements and GPU texture tiles.
+- Efficient and accurate invalidation of paint and raster.
+- Measuring
+[layout shift](https://web.dev/cls/) and
+[largest contentful paint](https://web.dev/lcp/) in Core Web Vitals.
+
+Every web document has four separate property trees: transform, clip, effect and scroll.(*)
+The transform tree represents CSS transforms and scrolling
+(a scroll transform is represented as a 2D transform matrix).
+The clip tree represents
+[overflow clips](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow).
+The effect tree represents all other visual effects—opacity, filters, masks,
+blend modes, and other kinds of clips such as clip-path.
+The scroll tree represents information about scrolling,
+such as how scrolls
+[chain](https://developer.mozilla.org/en-US/docs/Web/CSS/overscroll-behavior) together;
+it is needed to perform scrolling on the compositor thread.
+Each node in a property tree represents a scroll or visual effect applied by a DOM element.
+If it happens to have multiple effects,
+there may be more than one property tree node in each tree for the same element.
+
+The topology of each tree is like a sparse representation of the DOM.
+For example, if there are three DOM elements with overflow clips,
+then there will be three clip tree nodes,
+and the structure of the clip tree will follow
+the containing-block relationship between the overflow clips.
+There are also links between the trees.
+These links indicate the relative DOM hierarchy,
+and hence order of application, of the nodes.
+For example, if a transform on a DOM element is below another DOM element with a filter,
+then of course the transform applies before the filter.
+
+Each DOM element has a _property tree state_,
+which is a 4-tuple (transform, clip, effect, scroll)
+that indicates the nearest ancestor clip,
+transform and effect tree nodes that take effect on that element.
+This is very convenient, because with this information we know exactly the list of clips,
+transforms, and effects that apply to that element,
+and in which order.
+This tells us where it is on screen and how to draw it.
+
+{% Aside %}
+(*) There are four trees because scrolling only applies to the contained subtree.
+An element is contained by a scroller if it scrolls with it,
+and `position:absolute` and `position: fixed`
+elements often escape ancestor scrolling elements,
+in situations where the scrolling element is not in the containing block chain
+([containing block](https://developer.mozilla.org/en-US/docs/Web/CSS/Containing_block),
+recursively) of the fixed- or absolute-positioned element.
```

Comment:
>(containing block recursively)

What?

Reply:
It means what it says - find the containing block then then containing block of that containing block etc. Since this is a note and not central to the document, I think it's ok to be brief.

Reply:
That's not really a rule of writing. Whatever you put in has to be clear or it should be removed.

Reply:
I feel as if this would be clear to the intended audience for this post. Anyone reading this is likely to have a reasonable level of understanding and interest in the subject, these posts are long enough without expanding on everything.

Reply:
You're presuming that the solution will be long. Even if this is within the expertise of those reading, that's not the same issue as whether it's clear and you both may be too clear to really judge this. Please give me an explanation in this thread and maybe I'll have a simple way out.

Reply:
"The containing block chain for a DOM element is the [containing block](https://developer.mozilla.org/en-US/docs/Web/CSS/Containing_block), plus that containing block's containing block, and so on up to the root element"

Reply:
Yes! Thank you. Please put that in.